### ISSUE TITLE
[dataio] Should we retire dataio-pyshovel? (Trac #2054)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2046]
The only remaining feature to implement in dataio-shovel to make it feature complete with respect to dataio-shovel is pretty printing of objects.  There are at least options on the table.  Neither of which are technically difficult.  So it's not a question of if this will be implemented, but when.

Migrated from https://code.icecube.wisc.edu/ticket/2054
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:14:51", 
    "description": "The only remaining feature to implement in dataio-shovel to make it feature complete with respect to dataio-shovel is pretty printing of objects.  There are at least options on the table.  Neither of which are technically difficult.  So it's not a question of if this will be implemented, but when.", 
    "reporter": "olivas", 
    "cc": "david.schultz, cweaver", 
    "resolution": "fixed", 
    "_ts": "1550067291068818", 
    "component": "other", 
    "summary": "[dataio] Should we retire dataio-pyshovel?", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2017-07-26T16:14:14", 
    "milestone": "", 
    "owner": "cweaver", 
    "type": "cleanup"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170726T16:43:03: 
* david.schultz changed attachment from "" to "pretty_print"
* david.schultz commented:

> option 2 potential implementation

**Comment by icecube-trac** [comment]
Trac update at 20170726T16:43:37: david.schultz commented:

> Copying from olivas's email:
> 
> Option 1: Add a virtual method like print to I3FrameObject and add implementations for all derived classes as we go.
> Option 2: Keep all the printing code in dataio, separate from the class definitions.
> Option 3: <insert your idea here>

**Comment by icecube-trac** [comment]
Trac update at 20170726T17:40:55: olivas commented:

> On slack, I commented that I'm leaning towards Option 1.  Now I'm leaning away from it.  
> 
> The grounds right now are largely philosophical.  The purpose of I3FrameObject is to allow us to store objects of different types in the same container.  What does pretty printing have to do with that?  We might say the same about the serialization method, but serialization is integral to our frame objects, making it easier to render in downstream code, though nice, is not.
> 
> Adding a virtual print method to I3FrameObject could make our lives a little easier, since it's a fairly simple solution.  However, as Weaver mentioned on the call, this solution is "so very 90's" (paraphrasing).  The downside is based on slippery slopes and ending up with a class like ROOT's TObject.  Not likely, but let's try to avoid this if we can.

**Comment by icecube-trac** [comment]
Trac update at 20170726T18:36:23: cweaver commented:

> On the other hand, coding printing directly into dataio-shovel is something of an anti-pattern, where one can't just write a new object with an standard interface and have it work, but must also go add it to an obscure list in dataio-shovel. On a practical level dataio-shovel would also have to link against every library which defines a frame object type it wants to print, and would have to (in one translation unit) drag in the header for every such object.

**Comment by icecube-trac** [comment]
Trac update at 20170727T12:44:24: olivas commented:

> Replying to [comment:3 cweaver]:
> > On the other hand, coding printing directly into dataio-shovel is something of an anti-pattern, where one can't just write a new object with an standard interface and have it work, but must also go add it to an obscure list in dataio-shovel. On a practical level dataio-shovel would also have to link against every library which defines a frame object type it wants to print, and would have to (in one translation unit) drag in the header for every such object. 
> 
> Very true.

**Comment by icecube-trac** [comment]
Trac update at 20170728T18:12:00: jgonzalez commented:

> I tend to prefer an option similar to 1 but not quite.
> 
> I have found myself over and over writing a "string" method to pretty-print classes. The equivalent of __str__ in python, so instead of "print" it is "string" or "get_string" or whatever that returns a pretty string. Then this method can be called from streamers anywhere. I suppose shovel will use some sort of streamer, no?
> 
> I would not retire pyshovel.

**Comment by icecube-trac** [comment]
Trac update at 20171003T13:36:00: 
* olivas changed owner from "olivas" to "cweaver"
* olivas changed status from "new" to "assigned"
* olivas changed milestone from "October 2017" to "December 2017"

**Comment by icecube-trac** [comment]
Trac update at 20171003T13:41:11: olivas commented:

> We decided to go with option 1 in the second software parallel in Berlin and Chris kindly agreed to implement that.  Thanks Chris!

**Comment by icecube-trac** [comment]
Trac update at 20171003T14:54:19: david.schultz commented:

> Note that we also agreed in a 1-0 vote to remove pyshovel when this is done, providing a symlink to dataio-shovel.

**Comment by icecube-trac** [comment]
Trac update at 20171020T19:30:08: cweaver commented:

> After r158938 pretty printing in dataio-shovel is complete and working, as far as I am concerned.

**Comment by icecube-trac** [comment]
Trac update at 20171024T14:30:00: olivas commented:

> So it looks like we just need a symlink to dataio-shovel (called dataio-pyshovel), so the change is seamless to users.  Then we can officially kill dataio-pyshovel.

**Comment by icecube-trac** [comment]
Trac update at 20171024T16:34:19: 
* david.schultz changed resolution from "" to "fixed"
* david.schultz commented:

> In changeset 2999:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2999"
> retire dataio-pyshovel, making a symlink to dataio-shovel. closes #2046
> ```


* david.schultz changed status from "assigned" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T22:01:10: nwhitehorn commented:

> In changeset 169252:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="169252"
> retire dataio-pyshovel, making a symlink to dataio-shovel. closes #2046
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:14:51: 
* olivas changed milestone from "December 2017" to ""
* olivas commented:

> Milestone December 2017 deleted
